Language models

dmitriy-serdyuk opened the pull request:
Moved #5 here. Github cannot change destination =( isaacs/github#18

dmitriy-serdyuk committed:
Started LM integration into sequence generator

dmitriy-serdyuk committed:
Separate class for language model

dmitriy-serdyuk committed:
Added get dim for lookup table

dmitriy-serdyuk committed:
Made work sequence generator without LM

dmitriy-serdyuk committed:
Refactored LM in sequence generator

Unnecessary dict union

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -299,9 +316,16 @@ def generate(self, outputs, **kwargs):
+            lm_states = dict_subset(
+                kwargs, self._lm_state_names)
+            lm_arguments = self.language_model.apply_step(
+                outputs, as_dict=True, **dict_union(lm_states))
```
`dict_union(lm_states)`? Maybe just `lm_states`?

rizar commented on blocks/bricks/language_models.py:
```diff
@@ -0,0 +1,81 @@
+    See :class:`.Initializable` for initialization parameters.
+
+    """
+    def __init__(self, dim, transition=None, preaggregator=None, **kwargs):
```
I am not sure I understand what is `preaggregator`.

dmitriy-serdyuk replied:
In tests I use `LookupTable` as preaggregator.

rizar commented:
I am happy with language model support in `SequenceGenerator`, it can be merged. Instead of language model class `SequenceGenerator` should be used, I think. For your convenience I can merge this PR as is, but ideally I would like to wait until you use `SequenceGenerator` for the language model.

dmitriy-serdyuk commented:
There is still some work to be done for readout. Now, it just ignores the language model part.

rizar commented:
Well, not exactly. The language model states are passed to `readout`, thus it is sufficient to create a `Readout` with the right `source_names` in order for language model to be used. Does it make sense?

dmitriy-serdyuk commented:
Right, it just merges everything. In this case I'll use `SequenceGenerator` as a language model. 

Is there a way to apply some transformation before the transition? `outputs` is an integer matrix and we need to embed it before passing to a recurrency.

dmitriy-serdyuk commented:
I see, there is `LookupFeedback` for it, right?

rizar commented:
Exactly!

dmitriy-serdyuk commented on blocks/bricks/sequence_generators.py:
```diff
@@ -257,6 +266,13 @@ def cost_matrix(self, application_call, outputs, mask=None, **kwargs):
+        # Run the language model
+        if self.language_model:
+            lm_arguments = self.language_model.apply(
+                outputs, mask=mask, as_dict=True)
```
I cheated here, I shouldn't use outputs.

rizar replied:
Why?

dmitriy-serdyuk replied:
It is extra information which we don't have during testing time. 

I guess, I should use generated outputs here.

dmitriy-serdyuk replied:
Imagine, that we have so powerful LM that it can always reconstruct the output. In this case the cost is always 0 and the AM won't receive any gradient.

rizar replied:
I see what you mean and you are right, but wouldn't be sufficient to shift `lm_arguments` just like we do it with `states`?

dmitriy-serdyuk committed:
Used sequence generator for language model

Removed debugging stuff

dmitriy-serdyuk commented:
It should be it! 

I had to create another method which returns everything in order to extract language model states and outputs and reuse it in `cost_matrix` of a sequence generator.

rizar commented on blocks/bricks/language_models.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class LanguageModel(Feedforward, Initializable):
+    """The traditional recurrent transition.
```
No we don't need this file, do we?

rizar commented:
Does this PR contain the PR that changed default prototypes?

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -335,6 +377,10 @@ def get_dim(self, name):
         elif name == 'outputs':
             return self.readout.get_dim(name)
+        elif self.language_model and name in self._lm_state_names:
+            print '***************************************'
```
Remove me please!

dmitriy-serdyuk commented:
Oops, seems so.

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -270,8 +272,18 @@ def cost_matrix(self, application_call, outputs, mask=None, **kwargs):
+
+        # Run the language model
+        if self.language_model:
+            lm_arguments = self.language_model.evaluate(
```
Nit-picking, those are not arguments. You can call them results or deliveries, or anything else you want.

rizar replied:
`lm_states` would be probably the best one.

dmitriy-serdyuk replied:
Yes, agree.

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -270,8 +272,18 @@ def cost_matrix(self, application_call, outputs, mask=None, **kwargs):
+        if self.language_model:
+            lm_arguments = self.language_model.evaluate(
+                outputs=outputs, mask=mask, as_dict=True)
+            lm_arguments = dict_subset(lm_arguments, ['outputs'])
```
I think that a shift similar to the one in the line 267 should be applied to what currently is called `lm_arguments`. Do  you agree?

dmitriy-serdyuk committed:
Don't need language_models.py anymore

dmitriy-serdyuk committed:
Refactored lm_arguments

dmitriy-serdyuk commented:
I rebased and now it doesn't contain that PR, strange...

dmitriy-serdyuk commented:
I have a Theano Op which contains an FST and can output its states/outputs. It is quite easy to create a brick with it. Where should I place it?

rizar commented:
Please make a PR to a new file in `fully-neural-lvsr`. `lvsr/ops.py`?

rizar commented:
This PR is almost done, I only miss a shifting of LM states.

After merge I can upload a half-trained model to LISA file system, and you can try to attach a LM model to it.

dmitriy-serdyuk commented:
The trick is that `evaluate` shifts it! So I don't need to shift it second time.

rizar commented:
That's a nice idea, but don't you need `return_initial_states` for that?

dmitriy-serdyuk commented:
Right, I'll fix it.

dmitriy-serdyuk commented:
Actually, I don't need it. When I call `language_model.evaluate` [here](https://github.com/rizar/blocks/pull/6/files#diff-427a3bfe41a4e5724c82be4b2138143fR278), it applies transition with `return_initial_states` and shifts the feedback. Am I right?

dmitriy-serdyuk commented:
What is the easiest way to rename inputs/outputs of a brick? I have a problem that the language model has the same state names as the generator.

dmitriy-serdyuk committed:
Separate names for language model

rizar commented:
Your temporary solution with prefixes is a good one. I will read the code more carefully tomorrow.

rizar commented:
It seems to me you apply name mangling in `evaluate` but do not apply it in `generate`. This is rather confusing.. we should either apply it in  both places or nowhere, I think.

dmitriy-serdyuk committed:
Renaming trick in generate

dmitriy-serdyuk committed:
Fixed generate

dmitriy-serdyuk commented:
Finally, it works. Lots of renamings became quite ugly, though.

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -176,10 +185,17 @@ def _push_allocation_config(self):
+        for name in self.readout.source_names:
+            if name in transition_sources:
+                self.readout.source_dims.append(self.transition.get_dim(name))
+            elif (self.language_model and name.startswith('lm_') and
```
`name.startswith('lm_')` is redundant

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -299,12 +334,22 @@ def generate(self, outputs, **kwargs):
+                         in self.language_model.generate(
+                            outputs, as_dict=True, iterate=False,
+                            **lm_states).items()
+                         if name != 'costs'])
```
Would it work without this condition?

rizar replied:
I am confused. You get rid of costs here, but not of outputs, which are just as useless. I guess the right thing to do is to filter both, so that you could delete all these `lm_outputs` special cases below.

dmitriy-serdyuk replied:
No, it wouldn't because readout cannot accept it. But I still want to have outputs there because of FST transition, I return probabilities there as outputs. Maybe I should rethink it and create a special emitter for FSTs.

rizar replied:
Calling probabilities outputs is quite a hack. You can make the probability vector a state of the `FSTTransition`, it seems to me that it would solve all problems. Do you agree?

dmitriy-serdyuk replied:
Why is it a hack? It seems, that outputs is the most appropriate place to put the probabilities.

rizar replied:
It should be all considered in the context that this brick will be wrapped into `SequenceGenerator`. I will say a more a bit later.

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -322,29 +368,42 @@ def generate_delegate(self):
+        if name == 'outputs':
             return self.readout.get_dim(name)
+        if self.language_model:
+            if name.startswith('lm_') and name in self._lm_state_names + ['lm_outputs']:
```
The first condition seems redundant.

rizar commented:
Great! I would spend some time on polishing the code though, see my comments.

dmitriy-serdyuk committed:
Refactored LM in sequence generator

Small refactor

rizar commented on blocks/bricks/sequence_generators.py:
```diff
@@ -322,29 +367,42 @@ def generate_delegate(self):
-        elif name == 'outputs':
+        if name == 'outputs':
             return self.readout.get_dim(name)
+        if self.language_model:
```
Two ifs can be replaced with one with `and` operator.

dmitriy-serdyuk replied:
Sorry, what do you mean?

rizar replied:
`if self.language_model and name in self._lm_state_names`

rizar commented:
All right, the code is great! The place where we need in now however changed. We now use subtrees in fully-neural-lvsr, which in fact means that `libs/blocks/` folder of that repositories contains a copy of Blocks, which is once in a while synchronized with a particular branch on github. Could you please make a PR with this code to `libs/blocks/` in my `fully-neural-lvsr` repo?

dmitriy-serdyuk commented:
Should I just copy these files?

rizar commented:
@janchorowski , I need your help here. What's the right way to merge this PR to `libs/blocks/`?

dmitriy-serdyuk commented:
It also seems, that we don't have git subtree support in the lab (although, it is 1.7.11.7 version which should have this functionality).

rizar commented:
Jan is actually using a custom `stree` script.

janchorowski commented:
The easiest way to merge it is to send me commit ID, and I'll just apply it and push. 

The slightly longer is to download git-stree https://github.com/tdd/git-stree and add it to the path, then let it know about the stree configuration (from the main repo directory cat .gitconfigs/stree-config >> .git/config) and then do git stree pull libs/blocks.

janchorowski commented:
Anyway, the PR has to be in the LVSR branch first.

rizar commented:
I was just about to merge it, but it seems like it still contains commits of an already merged PR about default prototypes. Dima, you are sure that you rebased it?

rizar commented:
All right, sort out things with the extra commits and press the button, I think I have just given you merge rights.

dmitriy-serdyuk committed:
Fixed generation with language model and test

Conflicts:
	blocks/bricks/sequence_generators.py

Conflicts:
	blocks/bricks/sequence_generators.py

dmitriy-serdyuk commented:
OK, now I rebased it on the right branch, I merge.

dmitriy-serdyuk commented:
@janchorowski , so the commit is ca5436c

janchorowski commented:
OK, it is integrated in https://github.com/janchorowski/fully-neural-lvsr/commit/44c0a4a178a3f8463f7babf8d6cf814e3321f51e

Please let me know if all's good, I am new to the stree/subtree thing.

dmitriy-serdyuk commented:
I don't have access.

janchorowski commented:
OK, you should have now. Sorry

janchorowski commented:
Hi, actually I did the merge badly - somehow it has overwritten changes we made with @rizar to the blocks subtree. I will do it tomorrow morning properly. Right now I am too tired to make it right. @dmitriy-serdyuk please wait for my email that it is again done.

rizar commented:
I am starting to think that we should not have switched to subtrees. One has to relearn how to use all the git tools. E.g. I used to enter the submodule and see the list commits in it by `git log`, and now I don't know how to do this, because the log will contain also commits to other subtrees, and so on and so forth.

The thought that we can always very quickly switch back cheers me up.

janchorowski commented:
You mean we should not have switched? I don't know. Submodules are super hard when you want to make tiny changes to the submodules, because you need to change the config and everyone has to remember to sync it. And then yes, switching back is easy. Let's defer the discussoin to after the deadline, and just make all work in the fully-neural repo as opposed to its dependencies. 

And you can always issue a `git log .` to restrict the edits to a particular directory. History squashing in merging can also be disabled. We just need to learn it.

janchorowski commented:
OK, all is sorted out in https://github.com/janchorowski/fully-neural-lvsr/commit/d65f59381b8a3daccf88d371b391f41c91fbabf8

Sorry for the mess.
